bundles/org.eclipse.e4.ui.workbench.swt/src/org/eclipse/e4/ui/workbench/swt/internal/copy/ViewPatternFilter.java

```diff
@@ -42,20 +43,8 @@ public boolean isElementSelectable(Object element) {
 
 	@Override
 	protected boolean isLeafMatch(Viewer viewer, Object element) {
-		if (element instanceof String) {
-			return false;
-		}
```

Comment:
I think the checking the type of `element` must stay here and below.

Reply:
But why? The method is supposed to always return `false`, if `element` is a `String`.

Given that `String` is immutable, `element instanceof MPartDescriptor` is therefore also guaranteed to return `false` for `Strings`, where the second part of the expression is then skipped.

Previously, the check was only there for "performance", to avoid executing any unnecessary operations. But the same is still the case right now:

```
  // Method descriptor #21 (Lorg/eclipse/jface/viewers/Viewer;Ljava/lang/Object;)Z
  // Stack: 4, Locals: 4
  protected boolean isLeafMatch(org.eclipse.jface.viewers.Viewer viewer, java.lang.Object element);
     0  aload_2 [element]
     1  instanceof org.eclipse.e4.ui.model.application.descriptor.basic.MPartDescriptor [16]
     4  ifeq 64
     7  aload_2 [element]
     8  checkcast org.eclipse.e4.ui.model.application.descriptor.basic.MPartDescriptor [16]
    11  astore_3 [desc]
    12  iconst_2
    13  anewarray java.lang.String [22]
    16  dup
    17  iconst_0
    18  aload_3 [desc]
    19  invokeinterface org.eclipse.e4.ui.model.application.descriptor.basic.MPartDescriptor.getLabel() : java.lang.String [24] [nargs: 1]
    24  aastore
    25  dup
    26  iconst_1
    27  aload_3 [desc]
    28  invokeinterface org.eclipse.e4.ui.model.application.descriptor.basic.MPartDescriptor.getCategory() : java.lang.String [28] [nargs: 1]
    33  aastore
    34  invokestatic java.util.stream.Stream.of(java.lang.Object[]) : java.util.stream.Stream [31]
    37  aload_3 [desc]
    38  invokedynamic 0 apply(org.eclipse.e4.ui.model.application.descriptor.basic.MPartDescriptor) : java.util.function.Function [37]
    43  invokeinterface java.util.stream.Stream.map(java.util.function.Function) : java.util.stream.Stream [41] [nargs: 2]
    48  aload_0 [this]
    49  invokedynamic 1 test(org.eclipse.ui.internal.dialogs.ViewPatternFilter) : java.util.function.Predicate [45]
    54  invokeinterface java.util.stream.Stream.anyMatch(java.util.function.Predicate) : boolean [49] [nargs: 2]
    59  ifeq 64
    62  iconst_1
    63  ireturn
    64  iconst_0
    65  ireturn
```

First the runtime checks whether `element` is an MPartDescriptor. If not, it'll jump to offset 64 and return `false`. In total, 5 instructions are executed.

For comparision, this is the byte code with an explicit `instanceof String` check:

```
// Method descriptor #21 (Lorg/eclipse/jface/viewers/Viewer;Ljava/lang/Object;)Z
  // Stack: 4, Locals: 4
  protected boolean isLeafMatch(org.eclipse.jface.viewers.Viewer viewer, java.lang.Object element);
     0  aload_2 [element]
     1  instanceof java.lang.String [22]
     4  ifeq 9
     7  iconst_0
     8  ireturn
     9  aload_2 [element]
    10  instanceof org.eclipse.e4.ui.model.application.descriptor.basic.MPartDescriptor [16]
    13  ifeq 73
    16  aload_2 [element]
    17  checkcast org.eclipse.e4.ui.model.application.descriptor.basic.MPartDescriptor [16]
    20  astore_3 [desc]
    21  iconst_2
    22  anewarray java.lang.String [22]
    25  dup
    26  iconst_0
    27  aload_3 [desc]
    28  invokeinterface org.eclipse.e4.ui.model.application.descriptor.basic.MPartDescriptor.getLabel() : java.lang.String [24] [nargs: 1]
    33  aastore
    34  dup
    35  iconst_1
    36  aload_3 [desc]
    37  invokeinterface org.eclipse.e4.ui.model.application.descriptor.basic.MPartDescriptor.getCategory() : java.lang.String [28] [nargs: 1]
    42  aastore
    43  invokestatic java.util.stream.Stream.of(java.lang.Object[]) : java.util.stream.Stream [31]
    46  aload_3 [desc]
    47  invokedynamic 0 apply(org.eclipse.e4.ui.model.application.descriptor.basic.MPartDescriptor) : java.util.function.Function [37]
    52  invokeinterface java.util.stream.Stream.map(java.util.function.Function) : java.util.stream.Stream [41] [nargs: 2]
    57  aload_0 [this]
    58  invokedynamic 1 test(org.eclipse.ui.internal.dialogs.ViewPatternFilter) : java.util.function.Predicate [45]
    63  invokeinterface java.util.stream.Stream.anyMatch(java.util.function.Predicate) : boolean [49] [nargs: 2]
    68  ifeq 73
    71  iconst_1
    72  ireturn
    73  iconst_0
    74  ireturn
```

Where now the runtime checks whether `element` is of type `String`. If it isn't, it'll jump to offset 9, otherwise it also simply returns `false`.

So in both cases, it'll cost 5 operations.

Reply:
> But why? The method is supposed to always return `false`, if `element` is a `String`.

You're absolutely right. I'm sorry, I wasn't paying close enough attention and thought it would be a logic error (multi-tasking is difficult).

The current state is correct and nice and readable. 👍🏽